edgedb/retry.py

```diff
@@ -0,0 +1,163 @@
+import asyncio
+import random
+import time
+
+from . import errors
+from . import transaction as _transaction
+
+
+DEFAULT_MAX_ITERATIONS = 3
+
+
+def default_backoff(attempt):
+    return (2 ** attempt) * 0.1 + random.randrange(100) * 0.1
+
+
+class AsyncIOIteration(_transaction.AsyncIOTransaction):
+    def __init__(self, retry, owner, iteration):
+        super().__init__(owner)
+        self.__retry = retry
+        self.__iteration = iteration
+
+    async def start(self):
+        if not self._managed:
+            raise errors.InterfaceError(
+                "Only managed retriable transactions are supported. "
+                "Use `async with transaction:`"
+            )
+        await self._start(single_connect=self.__iteration != 0)
+
+    async def __aexit__(self, extype, ex, tb):
+        try:
+            await super().__aexit__(extype, ex, tb)
+        except errors.EdgeDBError as err:
```

Comment:
This method is quite different from `Iteration.__exit__`. Is that on purpose? I think they're supposed to look very much alike, no?

For example, in `Iteration.__exit__` you except a `errors.ClientConnectionClosedError`, and here you except on `errors.EdgeDBError`.

The latter logic is also different, here you just raise `err` if `ex is None`; there you have a bunch of code that checks tags and returns `True` or raises.

Reply:
Nice catch. I've simplified the logic in async case and missed sync case.